### PR TITLE
feat: User based and post to post recommendations

### PR DESCRIPTION
**Pull request by Mitcheljager**
This PR uses the awesome https://github.com/ankane/disco gem to add recommendations.

There are two types of recommendations;
 1. User recommendations. These are general recommendations for posts based on your favorites and the favorites of other users. These show on the homepage for users who have favorites and recommendations based on those favorites.
 2. Post to post recommendations. These are recommendations for a single post based on users who have favorited this and other posts. These show on post show pages only if the post has any favorites and has recommendations. **It might turn out that this creates too many db records, I'm really not sure. If this is the case I might end up removing these recommendations.**
 
Both of these recommendations are only based on what favorites they have in common, they aren't based on tags, titles, or whatever else. Because of this they stand out from "Similar codes" on post pages, which are purely based on tags.
 
The recommendations are generated via a rake task. This rake task will be configured to run every x amount of time. I'm currently thinking every hour, similar to setting hotness for posts. Alternatively we could generate them on a job every time a user favorites a post. However recommendations are generated in one big batch all together so this might be too intensive.

Only 3 recommendations are generated. In the future we could expand on this by creating a full blown recommendations page with many more recommendations, but for now this is a good start.

The recommendations use the default settings from the gem, if the recommendations end up being too inaccurate we can play with the `factors` param. The gem states it can vary wildly what is best for your project.

## User recommendations
image

## Post to post recommendations
image

### CODE REVIEW

**Comment by CactusPuppy** [comment]
Looks really cool! I'll give this a review later today.

File: app/views/posts/_user_recommendations.erb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<% cache [current_locale, current_user.recommended_posts] do %>
```
Comment (CactusPuppy) [inline app/views/posts/_user_recommendations.erb:1]: Does `current_user.recommended_posts` have a good equality check for the cache?

**Review by CactusPuppy — APPROVED** [review]
LGTM! One small comment, but other than that, I'm excited to bring this in.

File: app/views/posts/_user_recommendations.erb (hunk 1 above)
Comment (Mitcheljager) [reply app/views/posts/_user_recommendations.erb:1]: It should do! It is just an active records collection so it should work like any other

**Pull request merged**